src/kernl/utils/debugger.py

```diff
@@ -41,21 +40,32 @@ def __init__(self, grid: list[int], inputs: list[torch.Tensor], shuffle: bool =
 
         self.current_grid_position = None
         self.constexpr = Union[int, str]
-        previous_boundary = 0
-
-        range_tensor_dict = dict()
-        for t in inputs:
-            assert t.device.type == "cuda", print(f"Tensor {t} is not on cuda")
-            range_tensor_dict[(previous_boundary, previous_boundary + t.nelement())] = t
-            previous_boundary = previous_boundary + t.nelement()
-
-        self.range_tensor_dict = RangeKeyDict(range_tensor_dict)
-        self.tensor_ptr: dict[torch.Tensor, int] = {
-            tensor: range_ptrs[0] for range_ptrs, tensor in range_tensor_dict.items()
-        }
+        self.previous_boundary = 0
+        self.tensor_dict: dict[tuple(int, int), torch.Tensor] = dict()
+        self.range_tensor_dict: RangeKeyDict = dict()
+        self.tensor_ptr: dict[torch.Tensor, int] = dict()
         self.total_gm_read = 0
         self.total_gm_write = 0
 
+    def _add_input(self, tensor: torch.Tensor):
```

Comment:
to make code safer we could check existance in this method "_add_input_if_not_exists", so a method can't create inconsistent state + it will reduce number of methods